### DISCUSSION TITLE
Unescape JSON logs in Grafana for better UX

### DISCUSSION THREAD

**Discussion by AntonPetrov83** [grafana/grafana#46521] [Observability]
**Is your feature request related to a problem? Please describe.**
Log messages and especially multiline exception stack traces are hard to read when contain valid escaped JSON strings.

**Describe the solution you'd like**
I think it is expected that when we apply the `| json` formatter it should unescape JSON strings for better UX.

**Additional context**
Currently strings containing escaped `\"` are not very pleasant to read:
![image](https://user-images.githubusercontent.com/1997978/154408744-bd5322c6-a35e-43b2-b488-100d6fc9526b.png)

Exception stack traces are very hard to read because `\n` was not unescaped:
![image](https://user-images.githubusercontent.com/1997978/154408939-8559b731-ba7e-490e-beb4-aa65388e88f3.png)

**Comment by dannykopping** [comment]
@AntonPetrov83 I've transferred your issue to the Grafana repo, since Loki is just the database; the frontend work is all handled within Grafana.

**Comment by zuchka** [comment]
Thanks for creating this issue, @AntonPetrov83 

There is a long history with this issue, and I believe there have been several features added to support this behavior.

Can you please take a look at this pr, and especially the edge cases noted in the feedback?

Can you see how your logs differ from the situations where this should work?

https://github.com/grafana/grafana/pull/31352

**Comment by niklas-stegmann** [comment]
@zuchka maybe it's the "Logs"-Panel, because there is no option to enable "Escape newlines"

![image](https://user-images.githubusercontent.com/32418474/156885447-e669cc93-a899-4350-bda9-bd88be167e4b.png)

Is it planned to be added there too?

**Comment by pokab** [comment]
**Is your feature request related to a problem? Please describe.**
My problem seems related. However, instead of escape sequences (\n \t), only literal newline and tab characters are present in the detected field. I know this is definitely true because when using label_replace in the LogQL statement, I can replace "\n" or "\t" strings by any printable string. The replacement string shows up where the newlines or tabs should be. The original characters are not rendered at all.

**Describe the solution you'd like**
These newline characters and tabs are not rendered in the "log details" view (among the detected fields), which causes the details view to be unusable for viewing this type of preformatted text.

It would be nice to have (at least switchable) rendering of these characters.
Maybe even fixed-width characters.
Also, wrapping of too long lines should still work (ideally, switchable per-field).

**Additional context**
The content I'd like to view is a standard java stack trace, like this:
```
java.lang.Exception: example
	at somecode(file:line)
	at somecode(file:line)
	at somecode(file:line)
```

but really, anything with complex output should be viewed comfortably.

**Comment by ivanahuckova** [comment]
As this is a feature request to add **Escaping of incorrectly escaped newlines** to logs panel, I am going to change this to discussion. I agree that this would be a valuable feature and l'll bring this up during the next planning.

**Comment by chaudum** [comment]
The button to escape/unescape newlines for logs can cause confusion to users, because it shows up even when the special characters in the logs are correctly escaped, see https://github.com/grafana/loki/pull/7396#issuecomment-1284106785

**Reply by chaudum** [reply]
I'm not sure if that is actually a bug, because I don't know on which criteria the button shows up (or not).

**Comment by gelicia** [comment]
Hello, as you may have heard, we are [transitioning away from using discussions to discuss feature requests](https://github.com/grafana/grafana/discussions/73424). Due to the age and number of responses to this discussion, we are deciding to close it. If this is something you would like to see in Grafana, feel free to open an [issue](https://github.com/grafana/grafana/issues) so the discussion can continue. Thank you!
